diff -urN linux-3.14-rc1.orig/arch/x86/kernel/reboot.c linux-3.14-rc1.new/arch/x86/kernel/reboot.c
--- linux-3.14-rc1.orig/arch/x86/kernel/reboot.c	2014-02-07 13:43:01.050934031 +0100
+++ linux-3.14-rc1.new/arch/x86/kernel/reboot.c	2014-02-07 12:46:03.000000000 +0100
@@ -365,6 +365,14 @@
 			DMI_MATCH(DMI_PRODUCT_NAME, "Dell XPS710"),
 		},
 	},
+	{	/* Handle problems with rebooting on Dell Venue 8 Pro */
+		.callback = set_pci_reboot,
+		.ident = "Dell Venue 8 Pro",
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "DellInc."),
+			DMI_MATCH(DMI_PRODUCT_NAME, "Venue 8 Pro 5830"),
+		},
+	},
 
 	/* Hewlett-Packard */
 	{	/* Handle problems with rebooting on HP laptops */
